GTFS Trip-Modifications (#403)
google/transit: gtfs-realtime/spec/en/trip-modifications.md

## TripModifications

The modification is in effect on all of the listed service\_dates, until it is removed from the feed. On any given service date, a trip MUST NOT be assigned to more than one `TripModifications` object.

There MAY be multiple `TripModifications` for a given stop pattern. It may be desirable to split the trips into multiple modifications e.g. if the `propagated_modification_delay` changes significantly, over the course of the detour.

The trips created through GTFS-TripModifications modify and replace each specified `trip_id`, and don't create a copy or additional run.

The scheduled stop times of each replacement trip are created from those of the affected trip, by performing the changes listed in modifications. Stop sequences are assigned from 1 to n, increasing by 1 for each stop in the trip. A `TripUpdate` message must be provided to publish real-time arrival/departure times for the replacement trip.

## Linkage to TripUpdates

> * A TripUpdate SHOULD be provided using the replacement trip ID (`modifications_id` + `_` + `trip_id`). The TripUpdate MAY provide an ETA at the replacement stops.

Reviewer: See comment about rail shuttles here: https://github.com/google/transit/pull/403#issuecomment-1769581436. I think restricting modifications to have just a single replacement trip will prevent using this spec change for communicating useful info about rail shuttles.

> * A TripUpdate SHOULD be provided using the replacement trip ID (`modifications_id` + `_` + `trip_id`). The TripUpdate MAY provide an ETA at the replacement stops.

Author: SHOULD here refers to using the new modified trip_id. It can still be provided using the old unmodified trip_id but you loose the ability to refers to added stops in those updates. See also answer to [previous message](https://github.com/google/transit/pull/403#discussion_r1366048506)

> * A TripUpdate SHOULD be provided using the replacement trip ID (`modifications_id` + `_` + `trip_id`). The TripUpdate MAY provide an ETA at the replacement stops.

Author: When a trip is modified with a Modification, it should behave as-if the static GTFS was updated. So `SCHEDULED` is the right relationship to use.

* If no such TripUpdate is found, TripUpdates for the original `trip_id` will apply to the modified trip. In this case, no ETA is available at the replacement stops.

##  Modification

A `Modification` message replaces a span of n stop times (`num_stops_replaced`) from each affected trip starting at `start_stop_sequence`. The spans of the modifications MUST not overlap. Spans may not be contiguous; in this case the two modifications MUST be merged into one.  These stop times are replaced with a new stop time for each replacement stop described by `replacement_stops`.

> The sequence of `replacement_stops` may be of arbitrary length. For example, 3 stops could be replaced by 2, 4, or 0 stops as the situation may require.

Author: That's why we specify a new trip_id to be used in the TripUpdates, to be able to specify real time updates for those new stops. 

If the TU refers to the previous trip_id, then we use the reference from `stop_sequence` from the trip before modification. If the TU refers to the new modified trip,  then we use the reference from `stop_sequence` from the trip with modifications. I'll propose a clarifying text.

> The sequence of `replacement_stops` may be of arbitrary length. For example, 3 stops could be replaced by 2, 4, or 0 stops as the situation may require.

Author: Clarification here I pushed some clarification https://github.com/google/transit/pull/403/commits/9a0d6d5e9c790c2f22a88809a2993a9c2c14077d

# Trip Modifications

> A `TripModifications` message identifies a list of similar `trip_ids` from the (CSV) GTFS sharing the same stop pattern, which are all affected by particular modifications, such as a detour.

Reviewer: ```suggestion
A `TripModifications` message identifies a list of similar `trip_ids` from the (CSV) GTFS which are all affected by particular modifications, such as a detour.
```

The scheduled stop times of each replacement trip are created from those of the affected trip, by performing the changes listed in modifications. `stop_sequence` for all stop times are replaced by a new value of 1 to n, starting with 1 on the first stop_time and increasing by 1 for each stop in the trip. A `TripUpdate` message must be provided to publish real-time arrival/departure times for the replacement trip.

## Linkage to TripUpdates

* A TripUpdate SHOULD be provided using a `ModifiedTripSelector` inside the TripUpdate's `TripDescriptor`. 
    * When the TripUpdate refers to the replacement trip, the consumer should behave as if the static GTFS would have been modified with the TripModifications (e.g. `arrival_time`, `departure_time`, `stop_sequence`, `stop_id` on replacement stops)
    * When providing a `ModifiedTripSelector`, the other fields of the `TripDescriptor` MUST be left empty, to avoid confusion by consumer that aren't looking for the `ModifiedTripSelector` value. 

>     * TripUpdate feeds providing updates with `ModifiedTripSelector` SHOULD also include a TripUpdate targetting clients that don't about TripModifications. In other words, there should be two TripUpdates : one for clients with modified trips (with `TripModifications`) and one for clients with the originial unmodified GTFS (without `TripModifications`)

Reviewer 2: nit: "targetting clients that don't about TripModifications" - I think targeting is misspelled and I think there is a word missing between "don't about" (maybe it's "know"?).

>     * Providing a TripUpdate with the a `ModifiedTripSelector` is the only way to create predictions at replacement stops

Reviewer 3: ```suggestion
    * Providing a TripUpdate with a `ModifiedTripSelector` is the only way to create predictions at replacement stops
```

Author: Applied!

* If no such TripUpdate is found, TripUpdates for the original `trip_id` will apply to the modified trip. 
    * In this case, the static GTFS information used should be from the static GTFS before any TripModifications applied. 
    * Real time can be available to the common stops between the previous trip and the new modified trip, no ETA is available at the replacement stops.

##  Modification

> A `Modification` message replaces a span of n stop times (between `start_stop_selector` and `end_stop_selector`) from each affected trip starting at `start_stop_selector`. The spans of the modifications MUST not overlap. Spans may not be contiguous; in this case the two modifications MUST be merged into one.  These stop times are replaced with a new stop time for each replacement stop described by `replacement_stops`.

Reviewer: ```suggestion
A `Modification` message describes changes to each affected trip starting at `start_stop_selector`. There can be zero, one, or more than one stop time(s) replaced by a `Modification`. The spans of the modifications MUST not overlap. Spans may not be contiguous; in this case the two modifications MUST be merged into one.  These stop times are replaced with a new stop time for each replacement stop described by `replacement_stops`.
```

# Trip Modifications

> A `TripModifications` message identifies a list of similar `trip_ids` from the (CSV) GTFS sharing the same stop pattern, which are all affected by a particular modification, such as a detour.

Reviewer 4: ```suggestion
A `TripModifications` message identifies a list of similar `trip_ids` from the (CSV) GTFS sharing the same stop pattern, which are all affected by particular modifications, such as a detour.
```

## SLO: Service-level objective

> The frequency of data updates is expected to be approximately hourly (~24 times/day). Ingestion time may depend on the total number of affected trips. Consumers are expected to ingest a single TripModification within 5 minutes, and a feed with hundreds of detours within 20 minutes.

Reviewer 4: question: not a copy edit, but is the expectation that this data does not live within a `TripUpdates` feed which updates more frequently, or only that the data itself is not changing even if the feed itself is?

Author: We're not changing anything on the spec in regard on how feeds can contain multiple types of entities or not : ["A feed may, although not required to, combine entities of different types."](https://github.com/google/transit/pull/403/files#diff-90f282d10f519d763309d02a8161020994bcf89cfc6d45dabc6376e2b0c5492dR22) Swiftly merges TripModif with TripUpdates but we support any configuration. And yes, TripUpdates should update as frequently as needed and most likely will update more frequently than TripModifications. 

This section is to "manage expectations" that consumer can't ingest that type of data in seconds like we do for all other types of RT feeds. It's inspired by [Service-Changes](https://docs.google.com/document/d/1oPfQ6Xvui0g3xiy1pNiyu5cWFLhvrTVyvYsMVKGyGWM/edit#heading=h.dk9syw20pc6j), but I'm not sure how required it is.

## TripModifications

> The modification is in effect on all of the listed service\_dates, until it is removed from the feed. On any given service date, a trip MUST NOT be assigned to more than one `TripModifications` object.

Reviewer 4: ```suggestion
The `TripModifications` is in effect on all of the listed service\_dates, until it is removed from the feed. On any given service date, a trip MUST NOT be assigned to more than one `TripModifications` object.
```

> There MAY be multiple `TripModifications` for a given stop pattern. It may be desirable to split the trips into multiple modifications e.g. if the `propagated_modification_delay` changes significantly, over the course of the detour.

Reviewer 4: question: these two sentences appear to contradict each other.
- if it's not possible for a trip to have multiple `TripModification`s applied to it, is this saying that a single trip can only have one detour?
- if it is possible for a trip to have multiple `TripModification`s applied to it, how is the modified ID generated?

Additionally, https://github.com/google/transit/pull/403/files#diff-a1a14af3b4aed7d10a3b167231b52fdb483cf71b88f1655ed2e28bccd18800ceR1074 refers to multiple `TripModification`s applying to a single trip, so either this or that should be updated.

Author: There can be multiple `TripModifications` for the same stop pattern, but only one `TripModifications` per trip per service date. As mentioned, the delays values can potentially be different enough to warrant different `TripModifications` if needed. Or you could have two different detours for the same route with the same stop patter during one day (for example, a rush hour detour and a detour outside of rush hour). 

I think there might be a confusion between `TripModifications` and `Modification`. A `TripModifications` applies on a list of trip and [can contains multiple](https://github.com/google/transit/pull/403/files/9a0d6d5e9c790c2f22a88809a2993a9c2c14077d#diff-a1a14af3b4aed7d10a3b167231b52fdb483cf71b88f1655ed2e28bccd18800ceR1102) `Modification`. A `Modification` applies on a range of stop in the selected trips. The names are indeed very similar, which might be confusing.

Reviewer 4: Ah yes, that is a bit confusing, but it does make a lot of my questions go away. I think there are some other copy edits which might make it clearer that I'll go through and add.

> The trips created through GTFS-TripModifications modify and replace each specified `trip_id`, and don't create a copy or additional run. Modification are applied on the schedule information, like if a static GTFS (CSV) was modified. 

Reviewer 4: ```suggestion
The trips created through GTFS-TripModifications modify and replace each specified `trip_id`, and don't create a copy or additional run. Modifications are applied on the schedule information, like if a static GTFS (CSV) was modified. 
```

The scheduled stop times of each replacement trip are created from those of the affected trip, by performing the changes listed in modifications. `stop_sequence` for all stop times are replaced by a new value of 1 to n, starting with 1 on the first stop_time and increasing by 1 for each stop in the trip. The trip ID is also replaced by a replacement trip ID (`modifications_id` + `_` + `trip_id`). A `TripUpdate` message must be provided to publish real-time arrival/departure times for the replacement trip.

> ## Linkage to TripUpdates

Reviewer 2: I had a question about the linkage to TripUpdates.  I recognize there has been some back-and-forth on this in the comments, based on feedback from IBI Group and others, and I'm reflecting on the behavior of what happens for consumers who haven't updated yet to support trip modifications.

Apologies if I missed this up-thread, but I'm guessing you considered (and discarded) the option to just supply a TripUpdate with the existing `trip_id` that includes updates for:
* existing stops (no real change here)
* new stops in the trip (with possibly a new StopTimeUpdate. ScheduleRelationship enum value)
* skipped stops (either not mentioned at all or explicitly marking them SKIPPED)

When I think about consumers who might be slow to support trip modifications directly, having the separate trip update with a modified trip id would require some level of modification processing.  On the other hand, if the trip id was the same, these consumers might be able to more gracefully fallback to still apply stop time updates to their existing concept of the trip and hide the skipped stops (assuming they support SKIPPED).

I recognize `StopTimeUpdate.stop_sequence` indexing gets a bit murky here.

Author: The current text allows using the existing `trip_id` to affect existing stop and put SKIPPED on stops that are SKIPPED by the detour. That's what Swiftly is doing (and was doing before TripModifications too). 

On addressing new stops : I'm a bit worried that this would allow strange behaviour, validators will point that out as an error, etc. And yes, it wouldn't be possible to target stops with `stop_sequence`. You would also still need to ingest the new Stops. I feel having a light or partial implementation of TripModification would make more sense to me than try to allow everything in TripUpdates. 

One alternative is to use the experimental `stop_time_properties.assigned_stop_id` to assign new stops in real time for consumer that supports this and not TripModifications.

Reviewer 2: To be clear, I think I'm mostly fine with the high-level concept of a first-class "Trip Modification" itself (aka I'm not proposing to cram all this behavior into TripUpdates only).  But I'm keen to understand what TripUpdates would look like in a feed with modifications.

Swiftly has already mentioned that they are going down the path of publishing separate GTFS-Realtime feeds based on whether a consumer can handle trip modifications or not.  I know that's not the first instance of this happening with GTFS data in general, but it's still not a great precedent to set.  I'd rather just have one feed where a consumer who doesn't support TripModifications fully could still get some value out of the feed.  For example, we've already discussed why some consumers might not be able to support detours more generally, but most can realistically handle the skipped stops and stop-time updates for stops that haven't changed along the route.

But maybe this is all just hand-waving around my specific concern: I don't really like the "new trip id" = `modifications_id` + `_` + `trip_id` convention.  I'd rather put `modification_id` as a first-class field in `TripDescriptor` if you need that linkage.  But why is this modification id needed here at all?  If a TripUpdate specifies a trip id for a trip that has an active modification, then that linkage is already specified by the TripModification itself, yes?  Is there are world where a feed would include a TripUpdate for the non-modified trip and for the modified trip as the same time?

Reviewer 5: > Is there are world where a feed would include a TripUpdate for the non-modified trip and for the modified trip as the same time?

Swiftly does include TripUpdates for both the non-modified trip and the modified trip (when serving a request that wants modified trips). This is so there is no gap in real-time information between the time that we know about a modification, and the time that the consumer is able to ingest and apply it. To quite @mads14's [comment](https://github.com/google/transit/pull/403#issuecomment-1758406979):

> Meanwhile, we do still include all non-modified trip ids in our feed as well with the understanding that the consumer will replace those with the modified one when they read it, but given that it will take consumers some time (5 min or more) to read the trip modification details we didn't want to omit predictions for the normal trip id in the meantime.

Reviewer 2: Coming back to this point, I think the introduction of ModifiedTripSelector is an improvement over the id concatenation approach.  To the larger point, would it be worthwhile to add language below around the expectations for both producers and consumers for specifying updates for the original unmodified trip?  If we are arguing for backwards compatibility for existing clients, then I'd argue the unmodified trip should continue to be specified for the duration of the detour.  Thoughts?

Reviewer 6: Having the unmodified and modified trip in the trip updates could cause consumers that support predictions from trips with unidentified trip_ids, such as TransSee, to generate predictions for both if they aren't modified to handle it.

Author: @bdferris-v2 I didn't see that update from yesterday, but I agree that producer SHOULD provide both. I can update the language. 

@doconnoronca Did you review the changes since that original comment was posted? Specifically https://github.com/google/transit/pull/403/files#diff-abc780f2e1f0fd946abf169b0bb5d5ed06c33c695a028c7119043c0584eb7759R23

Reviewer 6: @gcamp Wouldn't that only work for consumers that have been updated to handle trip modifications?

Author: No, since there would be two trip updates, one with a "regular" `TripDescriptor` exactly like there is today (most of the time, with just a `trip_id`) and a second one with a `TripDescriptor` empty except for the `ModifiedTripSelector`. 

Clients that know about trip modification would use the update that has a `ModifiedTripSelector`. That update should be ignored by clients that don't know about trip modifications since the `TripDescriptor` would be empty otherwise. 

Clients that don't know about trip modifications would just use the update with "regular" `TripDescriptor`.

Reviewer 6: I see. Without a route_id the prediction couldn't be used without processing the ModifiedTripSelector.

* If no such TripUpdate is found, TripUpdates for the original `trip_id` will apply to the modified trip. 
    * In this case, the static GTFS information used should be from the static GTFS before any TripModifications applied. 
    * Real time can be available to the common stops between the previous trip and the new modified trip, no ETA is available at the replacement stops.

##  Modification

A `Modification` message replaces a span of n stop times (`num_stops_replaced`) from each affected trip starting at `start_stop_sequence`. The spans of the modifications MUST not overlap. Spans may not be contiguous; in this case the two modifications MUST be merged into one.  These stop times are replaced with a new stop time for each replacement stop described by `replacement_stops`.

The sequence of `replacement_stops` may be of arbitrary length. For example, 3 stops could be replaced by 2, 4, or 0 stops as the situation may require.

![](images/trip_modification.png)

_An example showing the effect of a modification on a particular trip. This modification may also be applied to several other trips._

![](images/propagated_delay.png)

> _Propagated detour delays affect all stops following the end of a modification. If a trip has multiple modifications, the delays are accumulated._

Reviewer 4: suggestion: This also mentions a trip with multiple modifications.

Author: Same as last comment

The trips created through GTFS-TripModifications modify and replace each specified `trip_id`, and don't create a copy or additional run. Modifications are applied on the schedule information, like if a static GTFS (CSV) was modified. 

The scheduled stop times of each replacement trip are created from those of the affected trip, by performing the changes listed in modifications. `stop_sequence` for all stop times are replaced by a new value of 1 to n, starting with 1 on the first stop_time and increasing by 1 for each stop in the trip. A `TripUpdate` message must be provided to publish real-time arrival/departure times for the replacement trip.

## Linkage to TripUpdates

* A TripUpdate SHOULD be provided using a `ModifiedTripSelector` inside the TripUpdate's `TripDescriptor`. 
    * When the TripUpdate refers to the replacement trip, the consumer should behave as if the static GTFS would have been modified with the TripModifications (e.g. `arrival_time`, `departure_time`, `stop_sequence`, `stop_id` on replacement stops)

>     * When providing a `ModifiedTripSelector`, the other fields of the `ScheduleRelationship` SHOULD be left empty, to avoid confusion by consumer that aren't looking for the `ModifiedTripSelector` value. 

Reviewer 2: Would it be an error to specify `trip_id` and `modified_trip` in the same TripDescriptor?

Author: I felt it was maybe a bit strong to disallow it entirely because it's not semantically wrong, but maybe it's better to enforce backward compatibility? It would be disallowed for `trip_id` but discouraged for other fields I think.  I don't have a strong opinion

Reviewer: I would disallow it for better backward compatibility unless there is a use case I'm not thinking of where providing both `trip_id` and `modified_trip` would be needed.

@gcamp Should "the other fields of the `ScheduleRelationship` SHOULD be left empty" be changed to "the other fields of the `TripDescriptor` SHOULD be left empty"? (And also change SHOULD to MUST if there is agreement on requiring this.)

![](images/propagated_delay.png)

_Propagated detour delays affect all stops following the end of a modification. If a trip has multiple modifications, the delays are accumulated._

## ReplacementStop

Each `ReplacementStop` message defines a stop that will now be visited by the trip, and optionally specifies the estimated travel time to the stop. The `ReplacementStop` message is used to construct the scheduled `stop_time` for the stop.

When `travel_time_to_stop` is specified, the `arrival_time` is calculated from a reference stop in the original trip, plus the offset in `travel_time_to_stop`. Otherwise, the `arrival_time` can be be interpolated based on the total duration of the modification in the original trip.

The `departure_time` always equals the `arrival_time`.

The optional fields of [`stop_times.txt`](https://github.com/google/transit/blob/master/gtfs/spec/en/reference.md#stop_timestxt) in the (CSV) GTFS specification are all set to their default values.

![](images/first_stop_reference.png)

_If a modification affects the first stop of the trip, that stop also serves as the reference stop of the modification._

> ## Shape

Reviewer 2: nit: I don't think there is value in including the Shape and Stop text here if they are going to replicate what's already specified in `[reference.md]`
